File: twoPointers/pySolutions/01-valid_palindrome-easy.py
'''
A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all 
non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters 
and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
'''
import re


def valid_palindrom(s):
    #remove special chars, spaces, and punctuation
    alpha_s = re.sub(r'[^A-Za-z0-9]+', '', s).lower()
    left = 0
    right = len(alpha_s) - 1

    while left <= right:
        if alpha_s[left] != alpha_s[right]:
            return False
        left += 1
        right -= 1

    return True


print(valid_palindrom("A man, a plan, a canal: Panama"))

File: twoPointers/jsSolutions/01-validPalindrome-easy.js
/* 
A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all 
non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters 
and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
*/


const validPalindrome = (str) => {
  let ns = str.replace(/[^a-z0-9]/gi, '')
  let left = 0
  let right = ns.length - 1

  while (left <= right) {
    if (ns[left].toLowerCase() !== ns[right].toLowerCase()) return false
    left++
    right--
  }

  return true
}

console.log(validPalindrome("A man, a plan, a canal: Panama"))
